Install udev rule as part of Debian pkg install

mdhorn opened the pull request:
Instead of requiring the user of the package to run a script, have
the udev rule installed as part of the package.

The udev will be installed as
    /lib/udev/rules.d/60-ros-<rosdistro>-kobuki-ftdi.rules
If more than one ROS distribution is installed on the same system,
the older distribution udev rules file will be used as the :=
operator is used.
See: https://www.kernel.org/pub/linux/utils/kernel/hotplug/udev/udev.html

mdhorn commented:
We had to build a new system for one of the Kobuki base turtlebot2 recently and were stuck one this for a while. We had done a system to system compare of the versions of ROS packages, but we had missed manually running the script.

stonier commented:
Can't drop the create scripts - kobuki is often built for systems which do not have debs. 

I am interested in getting it into the bloom process for the debs though, and a postinstall script might be exactly what is needed. Some questions though:

1. What part of this process is actually installing the file? The postinstall script does not seem to be doing this - it seems as though it is just active in clearing out old and reloading the rules.
2. How did you generate your deb? Locally in some custom way, or with bloom, or other?

mdhorn commented:
@stonier I'll update the PR to leave the manual script, but can I update the script to reference the new debian/udev file? Trying to avoid two copies of the udev rules file.

1. As part of building the debian package, the 'udev' file under the 'debian' directory will automatically be added to the package and installed under /lib/udev/rules.d as '60-<package_name>.rules as part of the standard debian packaging process.
See: http://manpages.ubuntu.com/manpages/xenial/man1/dh_installudev.1.html
Naming the file just 'udev', that file will be used for all package names. You can also instead have different udev rules per package name if for some reason the rules need to be different for kinetic versus indigo.

The 'postinst' script is needed to force the udev rules to reload after installation. I also added a check to inform the user that the manually installed file should be removed if present.

2. For testing, I was using `bloom-generate rosdebian` and then `fakeroot debian/rules binary`.

This is the same process steps I'm using for our librealsense driver which not only has to install udev rules, but also patch the kernel.
See: https://github.com/IntelRealSense/librealsense/tree/rosdebian/debian

mdhorn committed:
Install udev rule as part of Debian pkg install

Instead of requiring the user of the package to run a script, have
the udev rule installed as part of the package.

The udev will be installed as
    /lib/udev/rules.d/60-ros-<rosdistro>-kobuki-ftdi.rules
If more than one ROS distribution is installed on the same system,
the older distribution udev rules file will be used as the :=
operator is used.
See: https://www.kernel.org/pub/linux/utils/kernel/hotplug/udev/udev.html

Additional References:
    http://www.reactivated.net/writing_udev_rules.html
    http://manpages.ubuntu.com/manpages/xenial/man1/dh_installudev.1.html

mdhorn commented:
@stonier I've rebased this PR, and changed it to old add the postinst script and linked debin/udev to the existing 57-kobuki.rules file.

Let me know if there are questions or if you still think this is worth merging.

stonier commented:
Aye, it still is. Been in the middle of moving continents and a new job and without a Kobuki. It arrived last week though and you've probably already seen me in the middle of some updates - Yujin is hiring me to fix a few kinetic problems. Will look at this shortly.

stonier commented:
Ok, confirmed that it works for me with `bloom-generate rosdebian` and `fakeroot debian/rules binary`.

I have a doubt that this will work on the build farm though. I remember being pointed to by william to put post install scripts into the release repo. Couldn't test though - the usual `git-buildpackage` (now `gbp buildpackage`) invokation in the correct release repo branch is not working for me any longer.

So, going to merge this and test the resulting build farm deb. If that doesn't work, I'll try shifting this over to the bloom release repo.

stonier commented:
Great job, thanks! Coming in v0.7.8  :+1:
